src/api.rs

```diff
@@ -928,6 +985,75 @@ impl Client {
         }
     }
 
+    async fn obtain_sso_code(
+        &self,
+        sso_id: &str,
+    ) -> Result<(String, String, String)> {
+        let state: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(64)
+            .map(char::from)
+            .collect();
+
+        let sso_code_verifier: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(64)
+            .map(char::from)
+            .collect();
+
+        let mut hasher = Sha256::new();
+        hasher.update(sso_code_verifier.clone());
+        let code_challenge =
+            crate::base64::encode_url_safe_no_pad(hasher.finalize());
+
+        let Some(port) = (8065..8070)
+            .find(|&port| TcpListener::bind(("127.0.0.1", port)).is_ok())
+        else {
+            return Err(Error::FailedToFindFreePort {
+                range: "(8065..8070)".to_string(),
+            });
+        };
+
+        let listener = match TcpListener::bind(("127.0.0.1", port)) {
+            Ok(value) => value,
+            Err(e) => return Err(Error::CreateSSOCallbackServer { err: e }),
+        };
+
+        let callback_server =
+            async { start_sso_callback_server(&listener, state.as_str()) };
```

Comment:
this isn't going to work like you want - an async block that doesn't await anything isn't actually running asynchronously, it's just going to block the thread when it runs, which isn't what we want. you'll want to use async networking calls (from tokio, or a library which uses it) to make this work.